### PR TITLE
Remove ambiguity from the grammar in the specification

### PR DESCRIPTION
**Pull request by jhump**
There are a couple of constructions in the current grammar that are ambiguous.

For example, the current grammar could technically accept this:
```
message group {}
message foo {
    optional group g = 1;
}
```

But this is actually not allowed -- `group` is a valid name for the message definition on line 1, but it's not allowed as the type name of a field. In a field declaration it instead indicates a group field, which must have a body (instead of ending with a semicolon).

Another example that the current grammar could technically accept:
```
package stream;
message Foo
service FooService {
    rpc Do(stream.Foo) returns (stream.Foo);
}
```
The current grammar could interpret `stream.Foo` as the qualified name of a message type. However, the _correct_ way to parse this would be as a streaming RPC that uses `.Foo` as the request and response types. (In that case, the file is not valid because `.Foo` is not defined, though `.stream.Foo` is.)

There was some other text in the spec that attempted to explain constraints that are due to how `protoc` uses a predictive parser (so some tokens are not allowed in some contexts). I have removed all of that prose and codified in the grammar instead.

The example grammars are also updated. They are much more verbose since they do not support a "subtraction" operation in the grammar (which I added to the spec's EBNF variant purely to make the productions shorter and hopefully 🤞 easier to read).

### CODE REVIEW

**Comment by jhump** [comment]
cc @pkwarren, this is follow-on work from the cleanup I did in https://github.com/bufbuild/protocompile/pull/138

**Review by Alfus — APPROVED** [review]
Interesting parsing choices? in protoc :-)

**Comment by jhump** [comment]
> Interesting parsing choices?

For sure. A big part of why the language is so squirrely is because it never had a specification. "Whatever protoc does" is how the language was effectively specified, and many "choices" are simply unconsidered consequences of implementation details in protoc (particularly the bits about resolving [relative references](https://protobuf.com/docs/language-spec#relative-references) 🤮).

Our hope was to get this language spec accurate enough that it could eventually become the authority (maybe not this document, but an equivalent one formally published by Google). The intent was to move Protobuf towards being a language with an authoritative specification, instead of a language with an authoritative reference implementation.

File: docs/language-spec.md [hunk 1]
```diff
@@ -35,7 +35,11 @@ The syntax is specified using Extended Backus-Naur Form (EBNF):
 Production  = production_name "=" Expression "." .
 Expression  = Alternative { "|" Alternative } .
 Alternative = Term { Term } .
-Term        = production_name | literal [ "…" literal ] | Exclusion | Group | Option | Repetition .
+Term        = Reference | literal [ "…" literal ] |
+              Exclusion | Group | Option | Repetition .
+Reference   = production_name [ "-" Subtraction ] .
+Subtraction = Atom | "(" Atom { "|" Atom } ")" .
```
[the remark this answers is not in the document]
Comment (jhump) [reply docs/language-spec.md:41]: @Alfus, do you have thoughts about this part? Which do you think is better or more clear: `Subtraction` or `Subtrahend`?

[the remark this answers is not in the document]
Comment (Alfus) [reply docs/language-spec.md:41]: +1 to Subtraction
I have never seen Subtrahend before, thought it was a typo at first, and had to look it up

**Pull request merged**